Add tests for a ShoppingCart functionality

Chorniypistolet opened the pull request.

Chorniypistolet committed:
Docker HW v2

Chorniypistolet committed:
Spring Boot Testing

Chorniypistolet committed:
Corrected Spring Boot Testing

Chorniypistolet committed:
Corrected Spring Boot Testing v2

Chorniypistolet committed:
Spring Boot Testing : Add tests for a ShoppingCart functionality

Chorniypistolet committed:
Merge branch 'main' into develop

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controllerTests/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+
+    @BeforeEach
+    void setup() {
+        User mockUser = new User();
+        mockUser.setId(3L);
+        mockUser.setEmail("user1@example.com");
+
+        SecurityContextHolder.getContext().setAuthentication(
+                new UsernamePasswordAuthenticationToken(mockUser, null, List.of())
+        );
```
Can you relace it using WithMockUser annotation?

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controllerTests/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+package mate.academy.spring.boot.controllerTests;
```
```suggestion
package mate.academy.spring.boot.controller;
```
use the same package names (fix it everywhere)

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/repositoryTests/ShoppingCartRepositoryTest.java:
```diff
@@ -0,0 +1,86 @@
+        Long userId = 3L;
+        Optional<ShoppingCart> result = shoppingCartRepository.findByUserId(userId);
+        assertTrue(result.isPresent());
+        assertEquals(userId, result.get().getUser().getId());
```
Why do you compare only Ids? Can you compare objects?

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/repositoryTests/ShoppingCartRepositoryTest.java:
```diff
@@ -0,0 +1,86 @@
+        assertTrue(result.isEmpty());
+    }
+
+    @Test
+    @DisplayName("Test findByIdWithItems should return shopping cart with items for an existing ID")
+    @Sql(scripts = {
+            "classpath:database/add-users-to-users-table.sql",
+            "classpath:database/add-books-to-books-table.sql",
+            "classpath:database/add-shopping-carts-to-shopping-carts-table.sql",
+            "classpath:database/add-cart-items-to-cart-items-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {
+            "classpath:database/remove-cart-items-from-cart-items-table.sql",
+            "classpath:database/remove-shopping-carts-from-shopping-carts-table.sql",
+            "classpath:database/remove-books-from-books-table.sql",
+            "classpath:database/remove-users-from-users-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```
Can you move it above class?

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/serviceTests/ShoppingCartServiceTests.java:
```diff
@@ -0,0 +1,253 @@
+
+    @Test
+    @DisplayName("Should update cart item quantity and return updated ShoppingCartDto")
+    void testUpdateCartItemQuantity_WithCorrectQuantity_ShouldReturnUpdatedShoppingCartDto() {
```
better use Given-When-Then construction

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s improve your solution ;)

Chorniypistolet committed:
Fixed HW:
allocated blocks Given/When/Then in ShoppingCartServiceTests. In ShoppingCartControllerTest I don't use the annotation WithMockUser, but simulate the user manually, because using the annotation I don't get the user from which I take data for searching in my controller, because there I apply the annotation AuthenticationPrincipal

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    @BeforeEach
+    void setup() {
+        User mockUser = new User();
+        mockUser.setId(3L);
+        mockUser.setEmail("user1@example.com");
+
+        SecurityContextHolder.getContext().setAuthentication(
+                new UsernamePasswordAuthenticationToken(mockUser, null, List.of())
+        );
+    }
```
you can replace it with 
`@WithMockUser(username = "user1@example.com", roles = "USER")`

Chorniypistolet replied:
If I run a test with this annotation, I get an error: jakarta.servlet.ServletException: Request processing failed: java.lang.NullPointerException: Cannot invoke "mate.academy.spring.boot.model.User.getId()" because "user" is null

liliia-ponomarenko replied:
It should work correctly, so you can investigate this issue. (optional)

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+        expected.setCartItems(cartItemDtoSet);
+
+        MvcResult result = mockMvc.perform(post("/cart")
+                        .contentType("application/json")
```
```suggestion
                        .contentType(MediaType.APPLICATION_JSON)
```

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+        MvcResult result = mockMvc.perform(post("/cart")
+                        .contentType("application/json")
+                        .content(objectMapper.writeValueAsString(requestDto))
+                        .principal(() -> String.valueOf(userId)))
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+        ShoppingCartDto actual = objectMapper.readValue(result.getResponse().getContentAsString(), ShoppingCartDto.class);
+
+        assertNotNull(actual);
+        assertTrue(EqualsBuilder.reflectionEquals(expected, actual, "id"));
```
is it enough to compare only ids?

Chorniypistolet replied:
It seems to me that in this case, using EqualsBuilder, I compare the expected ShoppingCartDto with the actual one, excluding the parameter id, and additionally wrapping it in assertTrue
![image](https://github.com/user-attachments/assets/82f47993-61d8-46e6-8b61-15795aee4b6c)

liliia-ponomarenko reviewed (changes requested):
Good job! A few comments (but be attentive and fix it for each endpoint) ;)

Chorniypistolet committed:
Fixed HW v2

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/repository/ShoppingCartRepositoryTest.java:
```diff
@@ -0,0 +1,188 @@
+        book.setPrice(BigDecimal.valueOf(20));
+        return book;
+    }
+}
```
```suggestion
}

```

liliia-ponomarenko commented on src/test/java/mate/academy/spring/boot/controller/ShoppingCartControllerTest.java:
```diff
@@ -0,0 +1,309 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
```
Your test classes are not checking the checkstyle with the plugin, let's fix that

liliia-ponomarenko reviewed (changes requested):
Good job! But let’s  fix checkstyle issue before merging

Chorniypistolet committed:
Fixed HW v3: added checkstyle plugin
